cpp/src/arrow/compute/exec/exec_plan.cc

```diff
@@ -170,48 +165,409 @@ Status ExecPlan::Validate() { return ToDerived(this)->Validate(); }
+                                                 std::move(exprs));
+}
+
+struct SinkNode : ExecNode {
+  SinkNode(ExecNode* input, std::string label,
+           AsyncGenerator<util::optional<ExecBatch>>* generator)
+      : ExecNode(input->plan(), std::move(label), {input}, {"collected"}, {},
+                 /*num_outputs=*/0),
+        producer_(MakeProducer(generator)) {}
+
+  static PushGenerator<util::optional<ExecBatch>>::Producer MakeProducer(
+      AsyncGenerator<util::optional<ExecBatch>>* out_gen) {
+    PushGenerator<util::optional<ExecBatch>> gen;
+    auto out = gen.producer();
+    *out_gen = std::move(gen);
+    return out;
+  }
+
+  const char* kind_name() override { return "SinkNode"; }
+
+  Status StartProducing() override { return Status::OK(); }
+
+  // sink nodes have no outputs from which to feel backpressure
+  static void NoOutputs() { DCHECK(false) << "no outputs; this should never be called"; }
+  void ResumeProducing(ExecNode* output) override { NoOutputs(); }
+  void PauseProducing(ExecNode* output) override { NoOutputs(); }
+  void StopProducing(ExecNode* output) override { NoOutputs(); }
+
+  void StopProducing() override {
+    std::unique_lock<std::mutex> lock(mutex_);
+    StopProducingUnlocked();
+  }
+
+  void InputReceived(ExecNode* input, int seq_num, ExecBatch exec_batch) override {
+    std::unique_lock<std::mutex> lock(mutex_);
+    if (stopped_) return;
+
+    // TODO would be nice to factor this out in a ReorderQueue
+    if (seq_num <= static_cast<int>(received_batches_.size())) {
+      received_batches_.resize(seq_num + 1);
+      emitted_.resize(seq_num + 1, false);
+    }
+    received_batches_[seq_num] = std::move(exec_batch);
+    ++num_received_;
+
+    if (seq_num != num_emitted_) {
+      // Cannot emit yet as there is a hole at `num_emitted_`
+      DCHECK_GT(seq_num, num_emitted_);
+      return;
+    }
+
+    if (num_received_ == emit_stop_) {
+      StopProducingUnlocked();
+    }
+
+    // Emit batches in order as far as possible
+    // First collect these batches, then unlock before producing.
+    const auto seq_start = seq_num;
+    while (seq_num < static_cast<int>(emitted_.size()) && !emitted_[seq_num]) {
+      emitted_[seq_num] = true;
+      ++seq_num;
+    }
+    DCHECK_GT(seq_num, seq_start);
+    // By moving the values now, we make sure another thread won't emit the same values
+    // below
+    std::vector<ExecBatch> to_emit(
+        std::make_move_iterator(received_batches_.begin() + seq_start),
+        std::make_move_iterator(received_batches_.begin() + seq_num));
+
+    lock.unlock();
+    for (auto&& batch : to_emit) {
+      producer_.Push(std::move(batch));
+    }
+    lock.lock();
+
+    DCHECK_EQ(seq_start, num_emitted_);  // num_emitted_ wasn't bumped in the meantime
+    num_emitted_ = seq_num;
+  }
+
+  void ErrorReceived(ExecNode* input, Status error) override {
+    // XXX do we care about properly sequencing the error?
```

Comment:
I think we should not try and sequence the error in any fashion.  A similar question comes up with generators which is "when an error arrives, should we drain the queue of valid results first?  Or simply return the error as quickly as possible?"  I went with the "purge all valid results and return error as quickly as possible" approach.  However, I think @lidavidm and I disagreed on this point but neither of us felt too strongly about it.  A counterpoint that I think David brought up at the time was the case of someone trying to inspect or work with a broken dataset by using a limit operation to read as much as they could before the broken part.

Reply:
I'll remove this comment. (See also https://github.com/apache/arrow/pull/10397#discussion_r640799446 ). I think for now it's acceptable to consider an error in one node an error for all subsequent nodes, regardless of sequencing.